src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs

```diff
@@ -3,138 +3,591 @@
+            _handle = new MsQuicContextSafeHandle(handle, context, MsQuicApi.Api.ApiTable->StreamClose, SafeHandleType.Stream, connectionHandle);
+        }
+        catch
+        {
+            context.Free();
+            throw;
+        }
+
+        _defaultErrorCode = defaultErrorCode;
+
+        _canRead = true;
+        _canWrite = !flags.HasFlag(QUIC_STREAM_OPEN_FLAGS.UNIDIRECTIONAL);
+        if (!_canWrite)
+        {
+            _sendTcs.TrySetResult(final: true);
+        }
+        _id = (long)GetMsQuicParameter<ulong>(_handle, QUIC_PARAM_STREAM_ID);
+        _type = flags.HasFlag(QUIC_STREAM_OPEN_FLAGS.UNIDIRECTIONAL) ? QuicStreamType.Unidirectional : QuicStreamType.Bidirectional;
+        _startedTcs.TrySetResult();
+    }
+
+    internal ValueTask StartAsync(CancellationToken cancellationToken = default)
     {
-        private readonly MsQuicStream _provider;
+        ObjectDisposedException.ThrowIf(_disposed == 1, this);
 
-        internal QuicStream(MsQuicStream provider)
+        _startedTcs.TryInitialize(out ValueTask valueTask, this, cancellationToken);
         {
-            _provider = provider;
+            unsafe
+            {
+                int status = MsQuicApi.Api.ApiTable->StreamStart(
+                    _handle.QuicHandle,
+                    QUIC_STREAM_START_FLAGS.SHUTDOWN_ON_FAIL | QUIC_STREAM_START_FLAGS.INDICATE_PEER_ACCEPT);
+                if (StatusFailed(status))
+                {
+                    // TODO: aborted and the exception type
+                    _startedTcs.TrySetException(new MsQuicException(status));
+                }
+            }
         }
 
-        //
-        // Boilerplate implementation stuff
-        //
+        return valueTask;
+    }
+
+    public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+    {
+        ObjectDisposedException.ThrowIf(_disposed == 1, this);
+
+        if (!_canRead)
+        {
+            throw new InvalidOperationException(SR.net_quic_reading_notallowed);
+        }
+
+        if (NetEventSource.Log.IsEnabled())
+        {
+            NetEventSource.Info(this, $"{this} Stream reading into memory of '{buffer.Length}' bytes.");
+        }
+
+        if (_receiveTcs.IsCompleted)
+        {
+            // Special case exception type for pre-canceled token while we've already transitioned to a final state and don't need to abort read.
+            // It must happen before we try to get the value task, since the task source is versioned and each instance must be awaited.
+            cancellationToken.ThrowIfCancellationRequested();
+        }
+
+        int totalCopied = 0;
+        do
+        {
+            // Concurrent call, this one lost the race.
+            if (!_receiveTcs.TryGetValueTask(out ValueTask valueTask, this, cancellationToken))
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_io_invalidnestedcall, "read"));
```

Comment:
Why is this inside do...while ? One iteration can copy, and the following can throw?

Reply:
The loop can iterate at most twice. First, it tries to copy from the buffer, if it succeeds (data are already there) it will break out of the loop in `while (... totalCopied == 0);`. If no data are available in the buffer, the `await valueTask.ConfigureAwait(false);` will wait on the next `RECEIVE` event and once it arrives it will repeat the loop with copying, copy something (or learn about FIN flag) and break out of the loop.

The `valueTask` must be awaited each time it's retrieved, even if it's completed and it seems pointless - e.g.: `_receiveTcs.TrySetResult();`, since it's backed up by MRVTS which keeps "version" and the `await` will reset it (increment it).

Reply:
A comment like "this will loop at most twice" would be nice :)

Reply:
Done.

Reply:
I still don't understand how we would overcome the race when - between the first and the second iteration - a parallel read could come and "steal" the arrived data? What would happen to the first read? It seems like it could either throw or - would it start waiting on data to arrive again?